<a href="https://colab.research.google.com/github/Ankushsaini90/Banglore_house_price_predction/blob/main/price_prediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# data loading

In [109]:
# important libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import matplotlib
import seaborn as sns


In [110]:
# data set
df = pd.read_csv("/content/Bengaluru_House_Data.csv")
df.head()

area_type   availability                  location       size  \
0  Super built-up  Area         19-Dec  Electronic City Phase II      2 BHK   
1            Plot  Area  Ready To Move          Chikka Tirupathi  4 Bedroom   
2        Built-up  Area  Ready To Move               Uttarahalli      3 BHK   
3  Super built-up  Area  Ready To Move        Lingadheeranahalli      3 BHK   
4  Super built-up  Area  Ready To Move                  Kothanur      2 BHK   

   society total_sqft  bath  balcony   price  
0  Coomee        1056   2.0      1.0   39.07  
1  Theanmp       2600   5.0      3.0  120.00  
2      NaN       1440   2.0      3.0   62.00  
3  Soiewre       1521   3.0      1.0   95.00  
4      NaN       1200   2.0      1.0   51.00

In [111]:
# size of data
print("Rows: ", df.shape[0])
print("col: ", df.shape[1])

Rows:  13320
col:  9


In [112]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13320 entries, 0 to 13319
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   area_type     13320 non-null  object 
 1   availability  13320 non-null  object 
 2   location      13319 non-null  object 
 3   size          13304 non-null  object 
 4   society       7818 non-null   object 
 5   total_sqft    13320 non-null  object 
 6   bath          13247 non-null  float64
 7   balcony       12711 non-null  float64
 8   price         13320 non-null  float64
dtypes: float64(3), object(6)
memory usage: 936.7+ KB


In [113]:
df.describe(include = "all")

area_type   availability    location   size  society  \
count                  13320          13320       13319  13304     7818   
unique                     4             81        1305     31     2688   
top     Super built-up  Area  Ready To Move  Whitefield  2 BHK  GrrvaGr   
freq                    8790          10581         540   5199       80   
mean                     NaN            NaN         NaN    NaN      NaN   
std                      NaN            NaN         NaN    NaN      NaN   
min                      NaN            NaN         NaN    NaN      NaN   
25%                      NaN            NaN         NaN    NaN      NaN   
50%                      NaN            NaN         NaN    NaN      NaN   
75%                      NaN            NaN         NaN    NaN      NaN   
max                      NaN            NaN         NaN    NaN      NaN   

       total_sqft          bath       balcony         price  
count       13320  13247.000000  12711.000000  13320.000000  
unique       2117           NaN           NaN           NaN  
top          1200           NaN           NaN           NaN  
freq          843           NaN           NaN           NaN  
mean          NaN      2.692610      1.584376    112.565627  
std           NaN      1.341458      0.817263    148.971674  
min           NaN      1.000000      0.000000      8.000000  
25%           NaN      2.000000      1.000000     50.000000  
50%           NaN      2.000000      2.000000     72.000000  
75%           NaN      3.000000      2.000000    120.000000  
max           NaN     40.000000      3.000000   3600.000000

In [114]:
# null values
df.isnull().sum()

area_type          0
availability       0
location           1
size              16
society         5502
total_sqft         0
bath              73
balcony          609
price              0
dtype: int64

**We are assuming that columns like (area_type, availability, balcony, society) are not important to us**

In [115]:
# drop columns
df.drop(columns = ["area_type","availability","balcony","society"],inplace = True)
df.head()

location       size total_sqft  bath   price
0  Electronic City Phase II      2 BHK       1056   2.0   39.07
1          Chikka Tirupathi  4 Bedroom       2600   5.0  120.00
2               Uttarahalli      3 BHK       1440   2.0   62.00
3        Lingadheeranahalli      3 BHK       1521   3.0   95.00
4                  Kothanur      2 BHK       1200   2.0   51.00

In [116]:
# info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13320 entries, 0 to 13319
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   location    13319 non-null  object 
 1   size        13304 non-null  object 
 2   total_sqft  13320 non-null  object 
 3   bath        13247 non-null  float64
 4   price       13320 non-null  float64
dtypes: float64(2), object(3)
memory usage: 520.4+ KB


In [117]:
df.isnull().sum()

location       1
size          16
total_sqft     0
bath          73
price          0
dtype: int64

# Null values

In [118]:
# filling null value for location
df["location"]= df["location"].fillna("Whitefield")

In [119]:
# filling null value for size
df["size"] = df["size"].fillna("2 BHK")

In [120]:
# filling null value for bath
df["bath"] = df["bath"].fillna(df["bath"].median())

In [121]:
# info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13320 entries, 0 to 13319
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   location    13320 non-null  object 
 1   size        13320 non-null  object 
 2   total_sqft  13320 non-null  object 
 3   bath        13320 non-null  float64
 4   price       13320 non-null  float64
dtypes: float64(2), object(3)
memory usage: 520.4+ KB


# BHK

In [122]:
df.head()

location       size total_sqft  bath   price
0  Electronic City Phase II      2 BHK       1056   2.0   39.07
1          Chikka Tirupathi  4 Bedroom       2600   5.0  120.00
2               Uttarahalli      3 BHK       1440   2.0   62.00
3        Lingadheeranahalli      3 BHK       1521   3.0   95.00
4                  Kothanur      2 BHK       1200   2.0   51.00

In [123]:
df["bhk"] = df["size"].str.split().str.get(0).astype(int)

In [124]:
# drop size as we replace this columns with "BHK"
# df.drop(column = ["size"], inplace = True)
df.head()

location       size total_sqft  bath   price  bhk
0  Electronic City Phase II      2 BHK       1056   2.0   39.07    2
1          Chikka Tirupathi  4 Bedroom       2600   5.0  120.00    4
2               Uttarahalli      3 BHK       1440   2.0   62.00    3
3        Lingadheeranahalli      3 BHK       1521   3.0   95.00    3
4                  Kothanur      2 BHK       1200   2.0   51.00    2

In [125]:
df[df.bhk> 15]

location        size total_sqft  bath  price  bhk
1718       2Electronic City Phase II      27 BHK       8000  27.0  230.0   27
3379                  1Hanuman Nagar      19 BHK       2000  16.0  490.0   19
3609   Koramangala Industrial Layout      16 BHK      10000  16.0  550.0   16
4684                     Munnekollal  43 Bedroom       2400  40.0  660.0   43
11559                  1Kasavanhalli  18 Bedroom       1200  18.0  200.0   18

# square fit

In [126]:
df.head()

location       size total_sqft  bath   price  bhk
0  Electronic City Phase II      2 BHK       1056   2.0   39.07    2
1          Chikka Tirupathi  4 Bedroom       2600   5.0  120.00    4
2               Uttarahalli      3 BHK       1440   2.0   62.00    3
3        Lingadheeranahalli      3 BHK       1521   3.0   95.00    3
4                  Kothanur      2 BHK       1200   2.0   51.00    2

In [127]:
df["total_sqft"].unique()

array(['1056', '2600', '1440', ..., '1133 - 1384', '774', '4689'],
      dtype=object)

we have some values in range like "1133 - 1384". So we have to covert this into single value.

In [128]:
def convertRange(x):
  temp = x.split("-")
  if len(temp) == 2:
    return (float(temp[0]) + float(temp[1]))/2
  try:
    return float(x)
  except:
    return


In [129]:
df["total_sqft"]=df["total_sqft"].apply(convertRange)

In [130]:
df["total_sqft"].unique()

array([1056. , 2600. , 1440. , ..., 1258.5,  774. , 4689. ])

# price per sqft

In [131]:
df["price_per_sqft"] = df["price"]*100000 / df["total_sqft"]

In [132]:
df.head()

location       size  total_sqft  bath   price  bhk  \
0  Electronic City Phase II      2 BHK      1056.0   2.0   39.07    2   
1          Chikka Tirupathi  4 Bedroom      2600.0   5.0  120.00    4   
2               Uttarahalli      3 BHK      1440.0   2.0   62.00    3   
3        Lingadheeranahalli      3 BHK      1521.0   3.0   95.00    3   
4                  Kothanur      2 BHK      1200.0   2.0   51.00    2   

   price_per_sqft  
0     3699.810606  
1     4615.384615  
2     4305.555556  
3     6245.890861  
4     4250.000000

In [133]:
df.describe()

total_sqft          bath         price           bhk  price_per_sqft
count  13274.000000  13320.000000  13320.000000  13320.000000    1.327400e+04
mean    1559.626694      2.688814    112.565627      2.802778    7.907501e+03
std     1238.405258      1.338754    148.971674      1.294496    1.064296e+05
min        1.000000      1.000000      8.000000      1.000000    2.678298e+02
25%     1100.000000      2.000000     50.000000      2.000000    4.266865e+03
50%     1276.000000      2.000000     72.000000      3.000000    5.434306e+03
75%     1680.000000      3.000000    120.000000      3.000000    7.311746e+03
max    52272.000000     40.000000   3600.000000     43.000000    1.200000e+07

# outlier

In [134]:
# location
df["location"].value_counts()

location
Whitefield                        541
Sarjapur  Road                    399
Electronic City                   302
Kanakpura Road                    273
Thanisandra                       234
                                 ... 
Bapuji Layout                       1
1st Stage Radha Krishna Layout      1
BEML Layout 5th stage               1
singapura paradise                  1
Abshot Layout                       1
Name: count, Length: 1305, dtype: int64

we have some location count 1. so we put them together as "other" location.

we have 1305 locations

In [135]:
df["location"] = df["location"].apply(lambda x:x.strip())
location_count = df["location"].value_counts()

In [136]:
location_count

location
Whitefield                        542
Sarjapur  Road                    399
Electronic City                   304
Kanakpura Road                    273
Thanisandra                       237
                                 ... 
Bapuji Layout                       1
1st Stage Radha Krishna Layout      1
BEML Layout 5th stage               1
singapura paradise                  1
Abshot Layout                       1
Name: count, Length: 1294, dtype: int64

now after using Strip(remove space) we have 1294 rows.

In [137]:
location_less_10 = location_count[location_count <= 10]
location_less_10

location
Dairy Circle                      10
Nagappa Reddy Layout              10
Basapura                          10
1st Block Koramangala             10
Sector 1 HSR Layout               10
                                  ..
Bapuji Layout                      1
1st Stage Radha Krishna Layout     1
BEML Layout 5th stage              1
singapura paradise                 1
Abshot Layout                      1
Name: count, Length: 1053, dtype: int64

In [138]:
# location which are less than 10, will consider as other
df["location"] = df["location"].apply(lambda x: "other" if x in location_less_10 else x)

In [139]:
df["location"].value_counts()

location
other                 2885
Whitefield             542
Sarjapur  Road         399
Electronic City        304
Kanakpura Road         273
                      ... 
Nehru Nagar             11
Banjara Layout          11
LB Shastri Nagar        11
Pattandur Agrahara      11
Narayanapura            11
Name: count, Length: 242, dtype: int64

In [140]:
df.describe()

total_sqft          bath         price           bhk  price_per_sqft
count  13274.000000  13320.000000  13320.000000  13320.000000    1.327400e+04
mean    1559.626694      2.688814    112.565627      2.802778    7.907501e+03
std     1238.405258      1.338754    148.971674      1.294496    1.064296e+05
min        1.000000      1.000000      8.000000      1.000000    2.678298e+02
25%     1100.000000      2.000000     50.000000      2.000000    4.266865e+03
50%     1276.000000      2.000000     72.000000      3.000000    5.434306e+03
75%     1680.000000      3.000000    120.000000      3.000000    7.311746e+03
max    52272.000000     40.000000   3600.000000     43.000000    1.200000e+07

In [141]:
(df["total_sqft"]/df["bhk"]).describe()

count    13274.000000
mean       575.074878
std        388.205175
min          0.250000
25%        473.333333
50%        552.500000
75%        625.000000
max      26136.000000
dtype: float64

square_fit per bhk goes minimun to 0.25 which is not making sense so we consedir them as outlier.

In [142]:
df = df[((df["total_sqft"]/df["bhk"])>=300)]
df.describe()

total_sqft          bath         price           bhk  price_per_sqft
count  12530.000000  12530.000000  12530.000000  12530.000000    12530.000000
mean    1594.564544      2.559537    111.382401      2.650838     6303.979357
std     1261.271296      1.077938    152.077329      0.976678     4162.237981
min      300.000000      1.000000      8.440000      1.000000      267.829813
25%     1116.000000      2.000000     49.000000      2.000000     4210.526316
50%     1300.000000      2.000000     70.000000      3.000000     5294.117647
75%     1700.000000      3.000000    115.000000      3.000000     6916.666667
max    52272.000000     16.000000   3600.000000     16.000000   176470.588235

In [143]:
# cosdering pricr_per_sqft as (mean+std) or (mean-std)
def sqft_outlier_remover(df):
  df_output = pd.DataFrame()
  for key, subdf in df.groupby("location"):
    m = np.mean(subdf.price_per_sqft)
    st = np.std(subdf.price_per_sqft)
    gen_df = subdf[(subdf.price_per_sqft > (m-st)) & (subdf.price_per_sqft <= (m+st))]
    df_output = pd.concat([df_output,gen_df],ignore_index = True)
  return df_output
df = sqft_outlier_remover(df)
df.describe()

total_sqft          bath         price           bhk  price_per_sqft
count  10301.000000  10301.000000  10301.000000  10301.000000    10301.000000
mean    1508.440608      2.471702     91.286372      2.574896     5659.062876
std      880.694214      0.979449     86.342786      0.897649     2265.774749
min      300.000000      1.000000     10.000000      1.000000     1250.000000
25%     1110.000000      2.000000     49.000000      2.000000     4244.897959
50%     1286.000000      2.000000     67.000000      2.000000     5175.600739
75%     1650.000000      3.000000    100.000000      3.000000     6428.571429
max    30400.000000     16.000000   2200.000000     16.000000    24509.803922

In [144]:

def bhk_outlier_remover(df):
  exclude_indices = np.array([])
  for location, location_df in df.groupby("location"):
    bhk_stats = {
        "mean": np.mean(location_df.bhk),
        "std": np.std(location_df.bhk),
        "count": location_df.bhk.count()
    }

    for bhk, bhk_df in location_df.groupby("bhk"):
      stats = bhk_stats.get(bhk-1)
      if stats and stats["count"] > 5:
        exclude_indices = np.append(exclude_indices, bhk_df[bhk_df.price_per_sqft < (stats["mean"])].index.values)
  return df.drop(exclude_indices, axis = "index")



In [145]:
df = bhk_outlier_remover(df)
df

location       size  total_sqft  bath   price  bhk  \
0      1st Block Jayanagar      4 BHK      2850.0   4.0  428.00    4   
1      1st Block Jayanagar      3 BHK      1630.0   3.0  194.00    3   
2      1st Block Jayanagar      3 BHK      1875.0   2.0  235.00    3   
3      1st Block Jayanagar      3 BHK      1200.0   2.0  130.00    3   
4      1st Block Jayanagar      2 BHK      1235.0   2.0  148.00    2   
...                    ...        ...         ...   ...     ...  ...   
10296                other      2 BHK      1353.0   2.0  110.00    2   
10297                other  1 Bedroom       812.0   1.0   26.00    1   
10298                other      3 BHK      1440.0   2.0   63.93    3   
10299                other      2 BHK      1075.0   2.0   48.00    2   
10300                other      4 BHK      3600.0   5.0  400.00    4   

       price_per_sqft  
0        15017.543860  
1        11901.840491  
2        12533.333333  
3        10833.333333  
4        11983.805668  
...               ...  
10296     8130.081301  
10297     3201.970443  
10298     4439.583333  
10299     4465.116279  
10300    11111.111111  

[10301 rows x 7 columns]

In [146]:
df.shape

(10301, 7)

In [147]:
df.drop(columns = ["size", "price_per_sqft"],inplace = True)

# Cleaned Data

In [148]:
df.head()

location  total_sqft  bath  price  bhk
0  1st Block Jayanagar      2850.0   4.0  428.0    4
1  1st Block Jayanagar      1630.0   3.0  194.0    3
2  1st Block Jayanagar      1875.0   2.0  235.0    3
3  1st Block Jayanagar      1200.0   2.0  130.0    3
4  1st Block Jayanagar      1235.0   2.0  148.0    2

In [149]:
df.to_csv("clean_data.csv")

In [150]:
X = df.drop(columns=["price"])
y = df["price"]

# model

In [151]:
# importing model libraries
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression, Lasso, Ridge
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.compose import make_column_transformer
from sklearn.pipeline import make_pipeline
from sklearn.metrics import r2_score

In [152]:
# spliting
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size = 0.2, random_state=0)

In [153]:
print(X_train.shape)
print(X_test.shape)

(8240, 4)
(2061, 4)


## linear regression

In [154]:
col_trans = make_column_transformer(
    (OneHotEncoder(sparse_output=False), ["location"]),
        remainder="passthrough"
        )

In [155]:
scaler = StandardScaler()

In [156]:
pipeline = make_pipeline(col_trans, scaler, LinearRegression())

In [157]:
pipeline.fit(X_train,y_train)

/usr/local/lib/python3.10/dist-packages/sklearn/compose/_column_transformer.py:1623: FutureWarning: 
The format of the columns of the 'remainder' transformer in ColumnTransformer.transformers_ will change in version 1.7 to match the format of the other transformers.
At the moment the remainder columns are stored as indices (of type int). With the same ColumnTransformer configuration, in the future they will be stored as column names (of type str).
To use the new behavior now and suppress this warning, use ColumnTransformer(force_int_remainder_cols=False).

  warnings.warn(


Pipeline(steps=[('columntransformer',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('onehotencoder',
                                                  OneHotEncoder(sparse_output=False),
                                                  ['location'])])),
                ('standardscaler', StandardScaler()),
                ('linearregression', LinearRegression())])

In [158]:
y_pred_lr = pipeline.predict(X_test)

In [159]:
r2_score(y_test,y_pred_lr)

0.8294552236661241

## lasso

In [160]:
pipe = make_pipeline(col_trans, scaler, Lasso())

In [161]:
pipe.fit(X_train,y_train)

/usr/local/lib/python3.10/dist-packages/sklearn/compose/_column_transformer.py:1623: FutureWarning: 
The format of the columns of the 'remainder' transformer in ColumnTransformer.transformers_ will change in version 1.7 to match the format of the other transformers.
At the moment the remainder columns are stored as indices (of type int). With the same ColumnTransformer configuration, in the future they will be stored as column names (of type str).
To use the new behavior now and suppress this warning, use ColumnTransformer(force_int_remainder_cols=False).

  warnings.warn(


Pipeline(steps=[('columntransformer',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('onehotencoder',
                                                  OneHotEncoder(sparse_output=False),
                                                  ['location'])])),
                ('standardscaler', StandardScaler()), ('lasso', Lasso())])

In [162]:
y_pred_lasso = pipe.predict(X_test)

In [163]:
r2_score(y_test,y_pred_lasso )

0.8222119691869108

## ridge

In [164]:
pipe = make_pipeline(col_trans, scaler, Ridge())

In [165]:
pipe.fit(X_train,y_train)

/usr/local/lib/python3.10/dist-packages/sklearn/compose/_column_transformer.py:1623: FutureWarning: 
The format of the columns of the 'remainder' transformer in ColumnTransformer.transformers_ will change in version 1.7 to match the format of the other transformers.
At the moment the remainder columns are stored as indices (of type int). With the same ColumnTransformer configuration, in the future they will be stored as column names (of type str).
To use the new behavior now and suppress this warning, use ColumnTransformer(force_int_remainder_cols=False).

  warnings.warn(


Pipeline(steps=[('columntransformer',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('onehotencoder',
                                                  OneHotEncoder(sparse_output=False),
                                                  ['location'])])),
                ('standardscaler', StandardScaler()), ('ridge', Ridge())])

In [166]:
y_pred_ridge = pipeline.predict(X_test)

In [167]:
r2_score(y_test, y_pred_ridge)

0.8294552236661241

# result


In [168]:
print("Linear Regression: ", r2_score(y_test, y_pred_lr))
print("Linear Regression(Lasso): ", r2_score(y_test, y_pred_lasso))
print("Linear Regression(ridge): ", r2_score(y_test, y_pred_ridge))

Linear Regression:  0.8294552236661241
Linear Regression(Lasso):  0.8222119691869108
Linear Regression(ridge):  0.8294552236661241


In [169]:
import pickle

In [170]:
pickle.dump(pipe, open("Ridge_Model.pk1","wb"))